presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PropertyDerivations.java

```diff
@@ -181,7 +182,15 @@ public ActualProperties visitEnforceSingleRow(EnforceSingleRowNode node, List<Ac
         @Override
         public ActualProperties visitAssignUniqueId(AssignUniqueId node, List<ActualProperties> inputProperties)
         {
-            return Iterables.getOnlyElement(inputProperties);
+            ActualProperties properties = Iterables.getOnlyElement(inputProperties);
+            if (properties.getNodePartitioning().isPresent()) {
+                // preserve input (possibly preferred) partitioning
```

Comment:
@sopel39 @martint In case when input is partitioned somehow, I think we want to preserve that and *add* partitioning on `unique`, not just reserve the original partitioning. E.g. `partitioned_on(col1)` is weaker than `partitioned_on(col1) && partitioned_on(unique)`.

Reply:
I was just thinking about that. Unfortunately, it's not currently possible to express or reason about multiple independent partitionings currently.

One could be tempted to assume that if input is partitioned on `S` and this operation also guarantees partitioning on `unique`, then it's partitioned on `S` + `{unique}`. But that's not so. The inference for partitioning goes like this:

```partitioned_on({k}) => partitioned_on({k} + X)```

but,

```partitioned_on({k} + X) =/=> partitioned_on({k})``` for non-empty X

Reply:
@martint @sopel39  Indeed, it would be incorrect to report `partitioned_on({k, unique})`, but can't we report two separate partitioning properties, [partitioned_on({k}), partitioned_on(unique)], using `List<ActualProperties>`? I did notice that many `visitXXX` methods assume that `List<ActualProperties> inputProperties` contains exactly one element and `visitJoin` method expects exactly 2 entries.

What are the implications of not being able to add partitioned_on(unique) for pre-partitioned streams? How much does this limit the set of queries where streaming aggregation will be possible? Does it exclude all queries where probe side is source partitioned or is an intermediate result of a join or an aggregation?

Reply:
The list of actual properties is one object per child node. They need to align.

Reply:
The context of `AssignUniqueId` is that `unique` symbol produced by it is used in aggregation on all probe rows when correlated subquery is resolved. Any partitioning on probe rows will allow not to have exchange below such aggregation. However, with `AssugnUniqueId` we don't really want to break any natural/preferred partitioning of source nodes that would allow to avoid repartitioning of probe rows before `LeftJoin`.

Reply:
@sopel39 

> Any partitioning on probe rows will allow not to have [remote] exchange below such aggregation.

That's right (assuming you refer to remote exchanges). I didn't realize that. Thanks for pointing it out.